File: backend/app/schemas/delivery.py
from datetime import date, datetime
from typing import Literal

from app.schemas.common import CamelModel, RiskLevel

DeliveryStatus = Literal["EXPECTED", "AT_RISK", "DISPATCHED", "PARTIAL", "RECEIVED"]
DeliveryTimelineKey = Literal["PO_APPROVED", "PO_SENT", "ACKNOWLEDGED", "DISPATCHED", "DELIVERED"]


class DeliveryTimelineStep(CamelModel):
    key: DeliveryTimelineKey
    label: str
    completed: bool


class DeliveryExceptionSchema(CamelModel):
    id: str
    severity: RiskLevel
    message: str
    detected_at: datetime
    suggested_action_label: str | None = None


class Delivery(CamelModel):
    id: str
    po_id: str
    po_number: str
    supplier_id: str
    supplier_name: str
    material_name: str
    expected_date: date
    quantity_ordered: int
    quantity_received: int
    status: DeliveryStatus
    risk_level: RiskLevel
    timeline: list[DeliveryTimelineStep]
    exceptions: list[DeliveryExceptionSchema]
    last_supplier_communication_at: datetime | None = None

File: frontend/types/delivery.ts
import type { ISODate, ISODateTime, RiskLevel } from "@/types/common";

export type DeliveryStatus = "EXPECTED" | "AT_RISK" | "DISPATCHED" | "PARTIAL" | "RECEIVED";

export interface DeliveryTimelineStep {
  key: "PO_APPROVED" | "PO_SENT" | "ACKNOWLEDGED" | "DISPATCHED" | "DELIVERED";
  label: string;
  completed: boolean;
}

export interface DeliveryException {
  id: string;
  severity: RiskLevel;
  message: string;
  detectedAt: ISODateTime;
  suggestedActionLabel?: string;
}

export interface Delivery {
  id: string;
  poId: string;
  poNumber: string;
  supplierId: string;
  supplierName: string;
  materialName: string;
  expectedDate: ISODate;
  quantityOrdered: number;
  quantityReceived: number;
  status: DeliveryStatus;
  riskLevel: RiskLevel;
  timeline: DeliveryTimelineStep[];
  exceptions: DeliveryException[];
  lastSupplierCommunicationAt?: ISODateTime;
}

export interface GoodsReceiptRecord {
  id: string;
  deliveryId: string;
  quantityReceived: number;
  qualityAccepted: boolean;
  receivedAt: ISODateTime;
  recordedBy: string;
}
